xla/hlo/analysis/hlo_replication_analysis.cc

```diff
@@ -37,6 +37,81 @@ limitations under the License.
 #include "xla/xla_data.pb.h"
 
 namespace xla {
+namespace {
+// When cross_partition_spmd is true, returns the partition IDs of all
+// replica groups in which a given replica participates. Specfically, the k-th
+// element of the outermost vector in the returned data structure holds the
+// partition IDs converted from the global IDs in a collective's
+// replica_groups field for replica k.
+//
+// When cross_partition_spmd is false, returns the replica IDs of all
+// replica groups in which a given partition participates. Specfically, the k-th
+// element of the outermost vector in the returned data structure holds the
+// replica IDs converted from the global IDs in a collective's replica_groups
+// field for partition k.
+std::vector<std::vector<std::vector<int64_t>>> GroupsForReplicas(
+    absl::Span<const ReplicaGroup> groups, int64_t num_partitions,
+    int64_t replica_count, bool cross_partition_spmd) {
+  std::vector<std::vector<std::vector<int64_t>>> groups_for_replicas(
+      cross_partition_spmd ? replica_count : num_partitions);
+  int64_t num_replicas = cross_partition_spmd ? replica_count : num_partitions;
```

Comment:
Swap these two lines, and use `num_replicas` in the definition of `groups_for_replicas`